File: T2/EB_BDD_GR9/01.SOAP_JAVA_EUREKABANK_GR09/01.SOAP_JAVA_EUREKABANK_GR09/02.CLICON/BDD_SOAP_JAVA_CLICON_G09/src/main/java/ec/edu/monster/servicio/EurekaService.java
/*
 * Cliente SOAP para el servicio EurekaBank
 * Adaptado para consumir el servicio web SOAP
 */
package ec.edu.monster.servicio;

import ec.edu.monster.modelo.Movimiento;
import jakarta.xml.ws.Service;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import javax.xml.namespace.QName;

/**
 * Clase servicio que actúa como cliente SOAP para EurekaBank
 * @author Grupo Monster G09
 */
public class EurekaService {
    
    // URL del servicio SOAP (ajustar según tu configuración)
    private static final String WSDL_URL = "http://10.183.38.246:8080/WS_EUREKABANK_SERVICIO/EurekaService?wsdl";
    
    // Namespace del servicio
    private static final String TARGET_NAMESPACE = "http://servicios.monster.edu.ec/";
    private static final String SERVICE_NAME = "EurekaService";
    private static final String PORT_NAME = "EurekaServicePort";
    
    // Código de empleado por defecto para operaciones
    private static final String COD_EMP_DEFAULT = "0001";
    
    /**
     * Obtiene los movimientos de una cuenta
     * @param cuenta Número de cuenta
     * @return Lista de movimientos
     */
    public List<Movimiento> traerMovimientos(String cuenta) {
        List<Movimiento> lista = new ArrayList<>();
        
        try {
            // Crear el servicio usando la URL del WSDL
            URL url = new URL(WSDL_URL);
            QName serviceName = new QName(TARGET_NAMESPACE, SERVICE_NAME);
            QName portName = new QName(TARGET_NAMESPACE, PORT_NAME);
            
            Service service = Service.create(url, serviceName);
            
            // Obtener el port usando la interfaz
            EurekaServicePort port = service.getPort(portName, EurekaServicePort.class);
            
            // Invocar el método
            Movimiento[] movimientos = port.leerMovimientos(cuenta);
            
            // Convertir array a lista
            if (movimientos != null) {
                lista = Arrays.asList(movimientos);
            }
            
        } catch (Exception e) {
            System.err.println("Error al consultar movimientos: " + e.getMessage());
            e.printStackTrace();
        }
        
        return lista;
    }
    
    /**
     * Registra un depósito en una cuenta
     * @param cuenta Número de cuenta
     * @param importe Importe a depositar
     * @return 1 si fue exitoso, 0 en caso contrario
     */
    public int regDeposito(String cuenta, double importe) {
        try {
            // Crear el servicio usando la URL del WSDL
            URL url = new URL(WSDL_URL);
            QName serviceName = new QName(TARGET_NAMESPACE, SERVICE_NAME);
            QName portName = new QName(TARGET_NAMESPACE, PORT_NAME);
            
            Service service = Service.create(url, serviceName);
            
            // Obtener el port usando la interfaz
            EurekaServicePort port = service.getPort(portName, EurekaServicePort.class);
            
            // Invocar el método
            port.registrarDeposito(cuenta, importe, COD_EMP_DEFAULT);
            
            return 1;
            
        } catch (Exception e) {
            System.err.println("Error al registrar depósito: " + e.getMessage());
            e.printStackTrace();
            return 0;
        }
    }
    
    /**
     * Registra un retiro de una cuenta
     * @param cuenta Número de cuenta
     * @param importe Importe a retirar
     * @return 1 si fue exitoso, 0 en caso contrario
     */
    public int regRetiro(String cuenta, double importe) {
        try {
            // Crear el servicio usando la URL del WSDL
            URL url = new URL(WSDL_URL);
            QName serviceName = new QName(TARGET_NAMESPACE, SERVICE_NAME);
            QName portName = new QName(TARGET_NAMESPACE, PORT_NAME);
            
            Service service = Service.create(url, serviceName);
            
            // Obtener el port usando la interfaz
            EurekaServicePort port = service.getPort(portName, EurekaServicePort.class);
            
            // Invocar el método
            port.registrarRetiro(cuenta, importe, COD_EMP_DEFAULT);
            
            return 1;
            
        } catch (Exception e) {
            System.err.println("Error al registrar retiro: " + e.getMessage());
            e.printStackTrace();
            return 0;
        }
    }
    
    /**
     * Registra una transferencia entre cuentas
     * @param cuentaOrigen Cuenta origen
     * @param cuentaDestino Cuenta destino
     * @param importe Importe a transferir
     * @return 1 si fue exitoso, 0 en caso contrario
     */
    public int regTransferencia(String cuentaOrigen, String cuentaDestino, double importe) {
        try {
            // Crear el servicio usando la URL del WSDL
            URL url = new URL(WSDL_URL);
            QName serviceName = new QName(TARGET_NAMESPACE, SERVICE_NAME);
            QName portName = new QName(TARGET_NAMESPACE, PORT_NAME);
            
            Service service = Service.create(url, serviceName);
            
            // Obtener el port usando la interfaz
            EurekaServicePort port = service.getPort(portName, EurekaServicePort.class);
            
            // Invocar el método
            port.registrarTransferencia(cuentaOrigen, cuentaDestino, importe, COD_EMP_DEFAULT);
            
            return 1;
            
        } catch (Exception e) {
            System.err.println("Error al registrar transferencia: " + e.getMessage());
            e.printStackTrace();
            return 0;
        }
    }
}

File: T2/EB_BDD_GR9/02.SOAP_DOTNET_EUREKABANK_GR09/02.SOAP_DOTNET_EUREKABANK_GR09/02.CLIESC/BDD_SOAP_DOTNET_CLIESC_G09/BDD_SOAP_DOTNET_CLIESC_G09/ec/edu/monster/servicio/EurekaService.cs
using System;
using System.Collections.Generic;
using System.ServiceModel;
using System.ServiceModel.Channels;
using BDD_SOAP_DOTNET_CLIESC_G09.ec.edu.monster.modelo;

namespace BDD_SOAP_DOTNET_CLIESC_G09.ec.edu.monster.servicio
{
    /// <summary>
    /// Cliente SOAP para el servicio EurekaBank
    /// </summary>
    public class EurekaService
    {
        // URL del servicio SOAP (ajustar según tu configuración)
        // Por defecto IIS Express usa un puerto dinámico, ajustar según corresponda
        private static readonly string SERVICE_URL = "http://10.183.38.246:51641/WSEureka.svc";
        
        // Código de empleado por defecto para operaciones
        private static readonly string COD_EMP_DEFAULT = "0001";

        /// <summary>
        /// Obtiene los movimientos de una cuenta
        /// </summary>
        public List<Movimiento> TraerMovimientos(string cuenta)
        {
            ChannelFactory<IWSEureka> channelFactory = null;
            try
            {
                var binding = new BasicHttpBinding();
                binding.MaxReceivedMessageSize = 2147483647;
                binding.MaxBufferSize = 2147483647;
                binding.ReaderQuotas.MaxDepth = 2147483647;
                binding.ReaderQuotas.MaxStringContentLength = 2147483647;
                binding.ReaderQuotas.MaxArrayLength = 2147483647;
                binding.ReaderQuotas.MaxBytesPerRead = 2147483647;
                binding.ReaderQuotas.MaxNameTableCharCount = 2147483647;
                binding.Security.Mode = BasicHttpSecurityMode.None;

                var endpoint = new EndpointAddress(SERVICE_URL);
                channelFactory = new ChannelFactory<IWSEureka>(binding, endpoint);
                var client = channelFactory.CreateChannel();

                Movimiento[] movimientos = client.TraerMovimientos(cuenta);
                
                if (channelFactory.State == CommunicationState.Opened)
                {
                    channelFactory.Close();
                }

                if (movimientos != null)
                {
                    return new List<Movimiento>(movimientos);
                }
                return new List<Movimiento>();
            }
            catch (EndpointNotFoundException ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error al consultar movimientos: {ex.Message}");
                throw new Exception($"No se pudo conectar al servicio SOAP en {SERVICE_URL}. " +
                    $"Verifique que el servicio esté corriendo y que la URL sea correcta. " +
                    $"Error: {ex.Message}", ex);
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error al consultar movimientos: {ex.Message}");
                throw new Exception($"Error al consultar movimientos: {ex.Message}", ex);
            }
            finally
            {
                if (channelFactory != null && channelFactory.State != CommunicationState.Closed)
                {
                    try
                    {
                        channelFactory.Abort();
                    }
                    catch { }
                }
            }
        }

        /// <summary>
        /// Registra un depósito en una cuenta
        /// </summary>
        public int RegDeposito(string cuenta, double importe)
        {
            try
            {
                var binding = new BasicHttpBinding();
                var endpoint = new EndpointAddress(SERVICE_URL);
                var channelFactory = new ChannelFactory<IWSEureka>(binding, endpoint);
                var client = channelFactory.CreateChannel();

                string resultado = client.RegistrarDeposito(cuenta, importe, COD_EMP_DEFAULT);
                channelFactory.Close();

                // El servicio retorna "OK" o un mensaje de error
                return resultado.Contains("OK") || resultado.Contains("exitoso") ? 1 : 0;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error al registrar depósito: {ex.Message}");
                return 0;
            }
        }

        /// <summary>
        /// Registra un retiro de una cuenta
        /// </summary>
        public int RegRetiro(string cuenta, double importe)
        {
            try
            {
                var binding = new BasicHttpBinding();
                var endpoint = new EndpointAddress(SERVICE_URL);
                var channelFactory = new ChannelFactory<IWSEureka>(binding, endpoint);
                var client = channelFactory.CreateChannel();

                string resultado = client.RegistrarRetiro(cuenta, importe, COD_EMP_DEFAULT);
                channelFactory.Close();

                return resultado.Contains("OK") || resultado.Contains("exitoso") ? 1 : 0;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error al registrar retiro: {ex.Message}");
                return 0;
            }
        }

        /// <summary>
        /// Registra una transferencia entre cuentas
        /// </summary>
        public int RegTransferencia(string cuentaOrigen, string cuentaDestino, double importe)
        {
            try
            {
                var binding = new BasicHttpBinding();
                var endpoint = new EndpointAddress(SERVICE_URL);
                var channelFactory = new ChannelFactory<IWSEureka>(binding, endpoint);
                var client = channelFactory.CreateChannel();

                string resultado = client.RegistrarTransferencia(cuentaOrigen, cuentaDestino, importe, COD_EMP_DEFAULT);
                channelFactory.Close();

                return resultado.Contains("OK") || resultado.Contains("exitoso") ? 1 : 0;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error al registrar transferencia: {ex.Message}");
                return 0;
            }
        }

        /// <summary>
        /// Valida el ingreso de un usuario
        /// </summary>
        public string ValidarIngreso(string usuario, string password)
        {
            try
            {
                var binding = new BasicHttpBinding();
                var endpoint = new EndpointAddress(SERVICE_URL);
                var channelFactory = new ChannelFactory<IWSEureka>(binding, endpoint);
                var client = channelFactory.CreateChannel();

                string resultado = client.ValidarIngreso(usuario, password);
                channelFactory.Close();

                return resultado;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"Error al validar ingreso: {ex.Message}");
                return "Error";
            }
        }
    }

    /// <summary>
    /// Interfaz del servicio WCF (debe coincidir con IWSEureka del servidor)
    /// </summary>
    [ServiceContract(Namespace = "http://tempuri.org/")]
    public interface IWSEureka
    {
        [OperationContract]
        string ValidarIngreso(string usuario, string password);

        [OperationContract]
        string ProbarConexionBD();

        [OperationContract]
        Movimiento[] TraerMovimientos(string cuenta);

        [OperationContract]
        string RegistrarDeposito(string cuenta, double importe, string codEmp);

        [OperationContract]
        string RegistrarRetiro(string cuenta, double importe, string codEmp);

        [OperationContract]
        string RegistrarTransferencia(string cuentaOrigen, string cuentaDestino, double importe, string codEmp);

        [OperationContract]
        double TraerSaldoCuenta(string cuenta);
    }
}

